Need check for unconstrained type parameters

```rust
struct Foo<T>(T, bool);

impl<X, Y> Foo<X> {
    fn test() -> Y {
        123
    }
}

fn main() {
    let a = Foo::test();
}
```

When we don't use a generic parameter on the impl type it is unconstrained for substitution.

```
error[E0207]: the type parameter `Y` is not constrained by the impl trait, self type, or predicates
 --> src/main.rs:3:8
  |
3 | impl<X,Y> Foo<X> {
  |        ^ unconstrained type parameter
```

--- a/gcc/rust/typecheck/rust-hir-type-check-toplevel.h
+++ b/gcc/rust/typecheck/rust-hir-type-check-toplevel.h
@@ -211,7 +211,8 @@ public:
 	  }
       }
 
-    auto self = TypeCheckType::Resolve (impl_block.get_type ().get ());
+    auto self
+      = TypeCheckType::Resolve (impl_block.get_type ().get (), &substitutions);
     if (self == nullptr || self->get_kind () == TyTy::TypeKind::ERROR)
       {
 	rust_error_at (impl_block.get_locus (), "failed to resolve impl type");

--- a/gcc/rust/typecheck/rust-hir-type-check-type.h
+++ b/gcc/rust/typecheck/rust-hir-type-check-type.h
@@ -95,9 +95,11 @@ class TypeCheckType : public TypeCheckBase
   using Rust::Resolver::TypeCheckBase::visit;
 
 public:
-  static TyTy::BaseType *Resolve (HIR::Type *type)
+  static TyTy::BaseType *
+  Resolve (HIR::Type *type,
+	   std::vector<TyTy::SubstitutionParamMapping> *mappings = nullptr)
   {
-    TypeCheckType resolver;
+    TypeCheckType resolver (mappings);
     type->accept_vis (resolver);
 
     if (resolver.translated == nullptr)
@@ -186,6 +188,11 @@ public:
 		    translated
 		      = SubstMapper::Resolve (translated, path.get_locus (),
 					      &args);
+		    if (translated->get_kind () != TyTy::TypeKind::ERROR
+			&& mappings != nullptr)
+		      {
+			check_for_unconstrained (args.get_type_args ());
+		      }
 		  }
 		else
 		  {
@@ -241,8 +248,39 @@ public:
   }
 
 private:
-  TypeCheckType () : TypeCheckBase (), translated (nullptr) {}
+  TypeCheckType (std::vector<TyTy::SubstitutionParamMapping> *mappings)
+    : TypeCheckBase (), mappings (mappings), translated (nullptr)
+  {}
 
+  void
+  check_for_unconstrained (std::vector<std::unique_ptr<HIR::Type> > &type_args)
+  {
+    std::map<std::string, Location> param_location_map;
+    std::set<std::string> param_tys;
+    for (auto &mapping : *mappings)
+      {
+	std::string sym = mapping.get_param_ty ()->get_symbol ();
+	param_tys.insert (sym);
+	param_location_map[sym]
+	  = mapping.get_generic_param ()->get_locus_slow ();
+      }
+
+    std::set<std::string> args;
+    for (auto &arg : type_args)
+      args.insert (arg->as_string ());
+
+    for (auto &exp : param_tys)
+      {
+	bool used = args.find (exp) != args.end ();
+	if (!used)
+	  {
+	    Location locus = param_location_map.at (exp);
+	    rust_error_at (locus, "unconstrained type parameter");
+	  }
+      }
+  }
+
+  std::vector<TyTy::SubstitutionParamMapping> *mappings;
   TyTy::BaseType *translated;
 };
 

--- a/gcc/testsuite/rust.test/xfail_compile/unconstrained_type_param.rs
+++ b/gcc/testsuite/rust.test/xfail_compile/unconstrained_type_param.rs
@@ -1,0 +1,12 @@
+struct Foo<T>(T, bool);
+
+impl<X, Y> Foo<X> {
+    // { dg-error "unconstrained type parameter" "" { target { *-*-* } } .-1 }
+    fn test() -> Y {
+        123
+    }
+}
+
+fn main() {
+    let a = Foo::test();
+}

